### PR TITLE
[SCOUCLOUD] add config validation and generation

### PR DESCRIPTION
**Pull request by sevenzing**
Added module to parse user config (https://www.notion.so/blockscout/scoutcloud-service-for-autodeploy-0c4c981908b7486cb378448ef16cef04?pvs=4#8ba4d72fe5a54d5f8ecf3d968c06413b) to yaml config

### CODE REVIEW

File: scoutcloud/scoutcloud/src/logic/config/variables/rpc_url.rs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+async fn check_jsonrpc_health(provider: &Provider<Http>) -> Result<(), anyhow::Error> {
+    provider
+        .get_chainid()
+        .await
+        .map_err(|e| anyhow::anyhow!("failed to check health of `rpc_url`: {e}"))?;
+
+    Ok(())
+}
+
+async fn check_any_trace_method(provider: &Provider<Http>) -> Result<(), anyhow::Error> {
```
Comment (rimrakhimov) [inline scoutcloud/scoutcloud/src/logic/config/variables/rpc_url.rs:90]: Why do we need to check all tracing methods and why isn't the info which tracing method have worked is not used further?

File: scoutcloud/scoutcloud/src/logic/config/variables/rpc_url.rs [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+            }
+        };
+    }
+    Ok(None)
+}
+
+fn generate_possible_ws_urls(mut url: Url) -> Result<Vec<Url>, anyhow::Error> {
+    url.set_scheme("ws")
+        .map_err(|_| anyhow::anyhow!("failed to set scheme to ws"))?;
+    Ok(vec![url.join("ws").unwrap(), url.join("wss").unwrap(), url])
```
Comment (rimrakhimov) [inline scoutcloud/scoutcloud/src/logic/config/variables/rpc_url.rs:142]: Where is that websockets generation schema used? It seems a little strange for the websockets to have schema like `ws://node_url.com/ws`, `ws://node_url.com/wss`

File: scoutcloud/scoutcloud/src/logic/config/variables/rpc_url.rs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+            .map_err(Error::Internal)?
+        {
+            parsed.push((
+                ParsedVariableKey::BackendEnv("ETHEREUM_JSONRPC_WS_URL".to_string()),
+                serde_json::Value::String(ws_url.to_string()),
+            ));
+        } else {
+            tracing::warn!(
+                "no valid websocket url found for `rpc_url`, skipping websocket configuration"
+            );
```
Comment (rimrakhimov) [inline scoutcloud/scoutcloud/src/logic/config/variables/rpc_url.rs:76]: Just to make sure, not setting websocket url does not require to set any disabling env, does it?

It seems that blockscout uses `ws://localhost:8546` as a default value in that case, so interested if that will not cause any problem for blockscout startup

File: scoutcloud/scoutcloud/src/logic/config/variables/token_symbol.rs [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+#[derive(Debug, Serialize, Deserialize)]
+pub struct TokenSymbol(String);
+derive_display_from_serialize!(TokenSymbol);
+
+macros::custom_env_var!(
+    TokenSymbol,
+    String,
+    [
+        (config, "config.network.currency.symbol"),
+        (config, "config.network.currency.name")
```
Comment (rimrakhimov) [inline scoutcloud/scoutcloud/src/logic/config/variables/token_symbol.rs:17]: Is it that the currency symbol and name are the same for all instances?

File: scoutcloud/scoutcloud/src/logic/config/validated.rs [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+macro_rules! parse_config_vars {
+    ($config:ident, $context:ident, $validated_config:ident, $is_partial_config:ident, { $($var:ident),* $(,)? }) => {
+        paste::item! {
+            $({
+                let value: Option<_> = $config.[<$var:snake>].into();
+                let maybe_value = match ($is_partial_config, value) {
+                    (_, Some(value)) => Some(value),
+                    (false, None) => <variables::[<$var:snake>]::[<$var:camel>] as UserVariable<_>>::maybe_default(&$context),
```
Comment (rimrakhimov) [inline scoutcloud/scoutcloud/src/logic/config/validated.rs:22]: What will happen if config should be full, but some values which do not have default (e.g., chain_id) will not have a value explicitly set?

File: scoutcloud/scoutcloud/src/logic/config/variables/node_type.rs [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+derive_display_from_serialize!(NodeType);
+derive_fromstr_from_deserialize!(NodeType);
+
+macros::custom_env_var!(NodeType, String, backend, "NODE_TYPE", {
+    fn new(v: String, _config: &ConfigValidationContext) -> Result<Self, Error> {
+        Self::from_str(&v).map_err(|_| Error::Validation(format!("unknown node_type: '{}'", v)))
+    }
+
+    fn maybe_default(_context: &ConfigValidationContext) -> Option<String> {
+        Some("geth".to_string())
```
Comment (rimrakhimov) [inline scoutcloud/scoutcloud/src/logic/config/variables/node_type.rs:27]: In notion default node type is `ganache`. Shouldn't this value correspond to the one in notion description?

File: scoutcloud/scoutcloud/src/logic/config/variables/token_symbol.rs (hunk 4 above)
Comment (rimrakhimov) [reply scoutcloud/scoutcloud/src/logic/config/variables/token_symbol.rs:17]: Found in notion description

File: scoutcloud/scoutcloud/src/logic/config/variables/rpc_url.rs (hunk 2 above)
Comment (rimrakhimov) [reply scoutcloud/scoutcloud/src/logic/config/variables/rpc_url.rs:142]: Found in notion description

File: scoutcloud/scoutcloud/src/logic/config/variables/instance_url.rs [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    }
+}
+
+macros::custom_env_var!(
+    InstanceUrl,
+    String,
+    [
+        (config, "blockscout.ingress.hostname"),
+        (config, "frontend.ingress.hostname")
+    ],
```
[the remark this answers is not in the document]
Comment (sevenzing) [reply scoutcloud/scoutcloud/src/logic/config/variables/instance_url.rs:27]: yes. In `default.yaml` provided by devops teams, there was `&host` (that points to `blockscout.ingress.hostname) for `frontend.ingress.hostname` variable. we decided to remove `&host` variable and directly specify it in code

[the remark this answers is not in the document]
Comment (sevenzing) [reply scoutcloud/scoutcloud/src/logic/config/variables/instance_url.rs:27]: image

image

File: scoutcloud/scoutcloud/src/logic/config/variables/rpc_url.rs (hunk 3 above)
Comment (sevenzing) [reply scoutcloud/scoutcloud/src/logic/config/variables/rpc_url.rs:76]: I just asked backend team, it's fine to simply omit WS env

about default `ws://localhost:8546`: I didn't find it in backend code. maybe it's just example value from docs?

File: scoutcloud/scoutcloud/src/logic/config/variables/rpc_url.rs (hunk 1 above)
Comment (sevenzing) [reply scoutcloud/scoutcloud/src/logic/config/variables/rpc_url.rs:92]: this is temp solution. we check any trace method and if there is not found trace method, we disable internal txns indexer

**Pull request merged**